### DISCUSSION TITLE
Node type with fenced code

### DISCUSSION THREAD

**Discussion by slanden** [pdml-lang/pdml-lang.github.io#7] [Q&A] [answered]
The [Types section](https://pdml-lang.github.io/docs/introduction/index.html#ch__15) of the Introduction states that a `code` node wraps the inner code inside a pair of `"""`. What was the reason for not going with the single quote? I think it's semantically the same, plus you don't have to hold down "shift".

The only reason I didn't ask about the backtick is because I think some keyboards don't have those.

The same documentation also states that instead, you can prefix the closing `]` with the node name, in that case it was `code`. The need for this (or code fences) also arises for HTML's `<style>` and `<script>` tags. What are your thoughts on having a more general node name to encompass these cases as well?

E.g.
```
[escape (lang=css)
  .class {}
escape]
```

Or perhaps another name like "block" or "fence"

There's nothing wrong with having separate node names for each; My main thought is that people might see it on several nodes and think they need (or want) to write all nodes that way.

`[p some [b bolded b] text p]`

The idea was that if there's only one node name that can fit all the use cases for that alternate syntax, people will associate them together, and not with any other node.

However, that's just a thought - I have no idea.

A downside to that is you probably would end up having `style`, `code`, and `script` nodes anyway, unless attributes are relied on such as `lang` and a "display-only: true | false" or something - to account for both `code` and `script`. Having those node types AND the `escape` node would defeat the purpose of the `escape` node. In fact, I don't think it should exist if the others do.

None of this is a complaint, just thoughts.

**Comment by pdml-lang** [comment, marked as answer]
> The [Types section](https://pdml-lang.github.io/docs/introduction/index.html#ch__15) of the Introduction states that a code node wraps the inner code inside a pair of """. What was the reason for not going with the single quote?

The reason was to do the same as in most programming languages, where `"` is used for string literals (e.g. `"foo"`). In most programming languages the single quote is used for single character literals (e.g. `'f'`).

> The same documentation also states that instead, you can prefix the closing ] with the node name

Yes, but that's just a basic example in [Introduction to PDML](https://pdml-lang.github.io/docs/introduction/index.html#ch__15)

The rules that are actually applied for raw text nodes in PDML/PML are specified in chapter [raw_text](https://pdml-lang.github.io/docs/extensions/reference_manual/index.html#raw_text) of the [PDML Extensions Reference Manual](https://pdml-lang.github.io/docs/extensions/reference_manual/index.html).

As explained in that chapter, 3 different syntaxes are supported, and if we use the [Text Block Syntax](https://pdml-lang.github.io/docs/extensions/reference_manual/index.html#raw_text_block_syntax) then the closing node must be written on a separate line. We cannot write `[code foo code]`. We have to write:
```
[code
    foo
code]
```
Note: Only the [Delimited Text Syntax](https://pdml-lang.github.io/docs/extensions/reference_manual/index.html#raw_text_delimiter_syntax) (with fences) is currently officially supported in _PML_, and the delimiter character must be `"`, `=`, or `~`.

Currently, the following PML nodes are raw text nodes: `code`, `input`, `output`, `html`, and `table_data`. More nodes might be added in the future. For more information please refer to the official [PML Nodes Reference Manual](https://www.pml-lang.dev/docs/reference_manual/index.html).

**Comment by slanden** [comment]
> The reason was to do the same as in most programming languages

That makes sense.

>  In most programming languages the single quote is used for single character literals

True, and I smiled when JavaScript allowed me to be lazy. But, I don't have a strong opinion either way. Double quotes are totally fine. Thanks for the explanation.

> Currently, the following PML nodes are raw text nodes: code, input, output, html, and table_data. More nodes might be added in the future.

That's more of what I was concerned about; that some nodes have special syntax, but not all. I was thinking that either having only a single, general node for this type of syntax or every node, then it might avoid a possible nuance.

For example, one thing that bothers me about HTML is that not all tags have the same rules about their closing tags.
- Some require none (`<img>`)
- Some allow omitting it (`<li>`)
- Some require both (`<div>`)
- Or worse, special rules like ["you can omit the closing tag if certain tags follow, or if this other case" (`<p>`)](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/p)

Although it may be widely known by now, in my opinion it's not intuitive. For those that don't have a closing tag, front end tooling usually works around this by adding a '/' at the end (`<input />`), which is technically only correct for tags that have closing tags, that are also allowed to skip their closing tag [facepalm].

I'm not saying it will be a problem for people, or me; I kinda like it, but I'm also on the fence about it. My only concern is whether or not one day there will be a similar point of confusion as the HTML example.

**Reply by pml-lang** [reply]
> That's more of what I was concerned about; that some nodes have special syntax, but not all.

I see your point and I fully agree.

In the first version of PML, only the following syntax was supported:
```
[code
    int i = 1;
code]
```

In a later version, the following two syntaxes were also supported
```
[code
    ~~~
    int i = 1;
    ~~~
]
...
[code int i = 1;]
```
The initial syntax variation was kept only for backwards compatibility.
But this is indeed a bit confusing, not really needed (because the fences syntax is more powerful), and it makes parsing `raw_text` nodes more challenging (requires look-ahead!), and more difficult to support in tools such as editor plugins.
In the next PML version, the initial syntax (`[code ... code]`) will be "deprecated and removed in an upcoming version".
Hence, all nodes will end with `]`.

> one thing that bothers me about HTML is that not all tags have the same rules about their closing tags ... Or worse, special rules like "you can omit the closing tag if certain tags follow"

These are horrible rules. I promise that PDML/PML will never ever have rules like these. We need to keep it "as simple as possible, but not simpler".

**Reply by slanden** [reply]
You cleared everything up for me regarding the syntax, there's just a couple more things that are not quite clear to me regarding semantics..

Are code fences allowed in any PDML node, or just nodes with the name `t:raw_text`?
If the latter, is the PML `code` type just a `t:raw_text` node with semantics on top?

**Reply by pml-lang** [reply]
> Are code fences allowed in any PDML node, or just nodes with the name t:raw_text?

In a *PDML* document code fences are only supported in `t:raw_text` nodes, e.g.:
```
[any_node_name [t:raw_text
    ~~~
    first line
    ...
    last line
    ~~~
]]
```

> is the PML code type just a t:raw_text node with semantics on top?

Yes, in *PML*, nodes like `code`, `table_data`, `html` etc. are specified to be of PDML type `raw_text`. This is achieved by configuring the PDML parser, as explained in the list item _Configuration by node name_ in chapter [Type Nodes](https://pdml-lang.github.io/docs/extensions/user_manual/index.html#type_nodes). Therefore no explicit `t:raw_text` child node is needed. In PML you can simply write:

```
[code
    ~~~
    first line
    ...
    last line
    ~~~
]
```

**Comment by slanden** [comment]
> the initial syntax ([code ... code]) will be "deprecated and removed in an upcoming version".

I'm thinking about writing a VS Code extension for this since there doesn't seem to be one yet, so that's a good thing to note.

> We need to keep it "as simple as possible, but not simpler".

I agree completely!

**Reply by pml-lang** [reply]
> I'm thinking about writing a VS Code extension

That would be awesome !!!

@tajmone is working already on a [plugin for SublimeText](https://github.com/tajmone/Sublime-PML).

A plugin for the very popular VS Code editor is really needed too, and would certainly be very much appreciated by the community.
